File: typescript/src/RecordStore.ts
// =============================================================================
// vapor-idx — RecordStore.ts
// FIX: renamed private field to _schema and added public `get schema()` getter.
//      QueryEngine accessed schema via this.store['schema'] bracket notation
//      (a TypeScript private-access hack). The getter is the correct approach.
// =============================================================================

import {
  VaporRecord,
  VaporSchema,
  TypeDefinition,
  IndexStrategy,
  VaporError,
} from "./types.js";
import { validateRecordData } from "./SchemaValidator.js";
import { ExactIndex } from "./indexes/ExactIndex.js";
import { KeywordIndex } from "./indexes/KeywordIndex.js";
import { PrefixIndex } from "./indexes/PrefixIndex.js";
import { RangeIndex } from "./indexes/RangeIndex.js";

interface TypeIndexes {
  exact: ExactIndex;
  keyword: KeywordIndex;
  prefix: PrefixIndex;
  range: RangeIndex;
  strategies: Map<string, IndexStrategy>;
}

export class RecordStore {
  private readonly records: Map<string, VaporRecord> = new Map();
  private readonly byType: Map<string, Set<string>> = new Map();
  private readonly typeIndexes: Map<string, TypeIndexes> = new Map();
  private idCounter = 0;

  // _schema is private storage; `get schema()` exposes it read-only to engines.
  private readonly _schema: VaporSchema;

  constructor(schema: VaporSchema) {
    this._schema = schema;
    this.initialiseTypeIndexes();
  }

  get schema(): VaporSchema {
    return this._schema;
  }

  private initialiseTypeIndexes(): void {
    for (const [typeName, typeDef] of Object.entries(this._schema.types)) {
      this.typeIndexes.set(typeName, buildTypeIndexes(typeDef));
      this.byType.set(typeName, new Set());
    }
  }

  store(typeName: string, data: Record<string, unknown>): string {
    const typeDef = this._schema.types[typeName];
    if (typeDef === undefined) {
      throw new VaporError(
        `Unknown type "${typeName}". Declare it in the schema before storing records.`,
      );
    }
    validateRecordData(typeName, typeDef, data);
    const id = this.nextId();
    const now = Date.now();
    const record: VaporRecord = {
      id,
      type: typeName,
      data: Object.freeze({ ...data }),
      _createdAt: now,
      _updatedAt: now,
    };
    this.records.set(id, record);
    this.byType.get(typeName)!.add(id);
    this.indexRecord(typeName, id, data);
    return id;
  }

  update(id: string, partial: Record<string, unknown>): void {
    const record = this.records.get(id);
    if (record === undefined)
      throw new VaporError(`Record "${id}" does not exist.`);
    const typeName = record.type;
    const typeDef = this._schema.types[typeName]!;
    validateRecordData(typeName, typeDef, { ...record.data, ...partial });
    this.unindexFields(
      typeName,
      id,
      record.data as Record<string, unknown>,
      Object.keys(partial),
    );
    const updatedData = Object.freeze({ ...record.data, ...partial });
    this.records.set(id, {
      id,
      type: typeName,
      data: updatedData,
      _createdAt: record._createdAt,
      _updatedAt: Date.now(),
    });
    this.indexFields(
      typeName,
      id,
      updatedData as Record<string, unknown>,
      Object.keys(partial),
    );
  }

  delete(id: string): void {
    const record = this.records.get(id);
    if (record === undefined) return;
    this.unindexRecord(record.type, id, record.data as Record<string, unknown>);
    this.byType.get(record.type)?.delete(id);
    this.records.delete(id);
  }

  get(id: string): VaporRecord | null {
    return this.records.get(id) ?? null;
  }
  has(id: string): boolean {
    return this.records.has(id);
  }
  getAll(): VaporRecord[] {
    return [...this.records.values()];
  }

  getAllByType(typeName: string): VaporRecord[] {
    const ids = this.byType.get(typeName);
    if (ids === undefined) return [];
    return [...ids].map((id) => this.records.get(id)!).filter(Boolean);
  }

  getTypeIdSet(typeName: string): ReadonlySet<string> {
    return this.byType.get(typeName) ?? EMPTY_SET;
  }

  getIndexes(typeName: string): TypeIndexes | undefined {
    return this.typeIndexes.get(typeName);
  }

  getTypes(): string[] {
    return [...this.byType.keys()];
  }

  private indexRecord(
    typeName: string,
    id: string,
    data: Record<string, unknown>,
  ): void {
    this.indexFields(typeName, id, data, Object.keys(data));
  }

  private indexFields(
    typeName: string,
    id: string,
    data: Record<string, unknown>,
    fields: string[],
  ): void {
    const ti = this.typeIndexes.get(typeName)!;
    for (const fieldName of fields) {
      const value = data[fieldName];
      const strategy = ti.strategies.get(fieldName);
      if (strategy === undefined || strategy === "none") continue;
      switch (strategy) {
        case "exact":
          ti.exact.add(fieldName, value, id);
          break;
        case "keyword":
          ti.keyword.add(fieldName, value, id);
          break;
        case "prefix":
          ti.prefix.add(fieldName, value, id);
          break;
        case "range":
          ti.range.add(fieldName, value, id);
          break;
      }
    }
  }

  private unindexRecord(
    typeName: string,
    id: string,
    data: Record<string, unknown>,
  ): void {
    this.unindexFields(typeName, id, data, Object.keys(data));
  }

  private unindexFields(
    typeName: string,
    id: string,
    data: Record<string, unknown>,
    fields: string[],
  ): void {
    const ti = this.typeIndexes.get(typeName)!;
    for (const fieldName of fields) {
      const value = data[fieldName];
      const strategy = ti.strategies.get(fieldName);
      if (strategy === undefined || strategy === "none") continue;
      switch (strategy) {
        case "exact":
          ti.exact.remove(fieldName, value, id);
          break;
        case "keyword":
          ti.keyword.remove(id);
          break;
        case "prefix":
          ti.prefix.remove(fieldName, value, id);
          break;
        case "range":
          ti.range.remove(fieldName, value, id);
          break;
      }
    }
  }

  get totalRecords(): number {
    return this.records.size;
  }

  get recordCountsByType(): Record<string, number> {
    const result: Record<string, number> = {};
    for (const [type, ids] of this.byType) result[type] = ids.size;
    return result;
  }

  get indexStats() {
    let exactEntries = 0,
      keywordTokens = 0,
      prefixNodes = 0,
      rangeEntries = 0;
    for (const ti of this.typeIndexes.values()) {
      exactEntries += ti.exact.entryCount;
      keywordTokens += ti.keyword.tokenCount;
      prefixNodes += ti.prefix.nodeCount;
      rangeEntries += ti.range.entryCount;
    }
    return { exactEntries, keywordTokens, prefixNodes, rangeEntries };
  }

  clear(): void {
    this.records.clear();
    for (const ti of this.typeIndexes.values()) {
      ti.exact.clear();
      ti.keyword.clear();
      ti.prefix.clear();
      ti.range.clear();
    }
    for (const ids of this.byType.values()) ids.clear();
  }

  private nextId(): string {
    return `vpr_${Date.now().toString(36)}_${(++this.idCounter).toString(36)}`;
  }
}

function buildTypeIndexes(typeDef: TypeDefinition): TypeIndexes {
  const strategies = new Map<string, IndexStrategy>();
  for (const [fieldName, fieldDef] of Object.entries(typeDef.fields)) {
    strategies.set(fieldName, fieldDef.index);
  }
  return {
    exact: new ExactIndex(),
    keyword: new KeywordIndex(),
    prefix: new PrefixIndex(),
    range: new RangeIndex(),
    strategies,
  };
}

const EMPTY_SET: ReadonlySet<string> = new Set();

File: python/vapor_idx/record_store.py
# =============================================================================
# vapor-idx — record_store.py  (v2.0 — adds update_where bulk conditional update)
# All existing methods are unchanged. Only update_where() is new.
# =============================================================================

from __future__ import annotations
import time
from dataclasses import dataclass
from typing import Any

from .types import VaporRecord, VaporSchema, IndexStrategy, VaporError
from .schema import validate_record_data
from .indexes import ExactIndex, KeywordIndex, PrefixIndex, RangeIndex


@dataclass
class _TypeIndexes:
    exact:      ExactIndex
    keyword:    KeywordIndex
    prefix:     PrefixIndex
    range_:     RangeIndex
    strategies: dict[str, IndexStrategy]


class RecordStore:
    def __init__(self, schema: VaporSchema) -> None:
        self._schema:    VaporSchema              = schema
        self._records:   dict[str, VaporRecord]   = {}
        self._by_type:   dict[str, set[str]]      = {}
        self._type_idxs: dict[str, _TypeIndexes]  = {}
        self._counter:   int                      = 0
        self._init_indexes()

    @property
    def schema(self) -> VaporSchema:
        """Public read-only access to the schema for QueryEngine and other engines."""
        return self._schema

    def _init_indexes(self) -> None:
        for type_name, type_def in self._schema.types.items():
            strats: dict[str, IndexStrategy] = {
                fn: fd.index for fn, fd in type_def.fields.items()
            }
            self._type_idxs[type_name] = _TypeIndexes(
                exact=ExactIndex(), keyword=KeywordIndex(),
                prefix=PrefixIndex(), range_=RangeIndex(),
                strategies=strats,
            )
            self._by_type[type_name] = set()

    def store(self, type_name: str, data: dict[str, Any]) -> str:
        type_def = self._schema.types.get(type_name)
        if type_def is None:
            raise VaporError(f'Unknown type "{type_name}". Declare it in the schema before storing records.')
        validate_record_data(type_name, type_def, data)
        record_id = self._next_id()
        now       = int(time.time() * 1000)
        record    = VaporRecord(id=record_id, type=type_name, data=dict(data), _created_at=now, _updated_at=now)
        self._records[record_id] = record
        self._by_type[type_name].add(record_id)
        self._index_record(type_name, record_id, data)
        return record_id

    def update(self, record_id: str, partial: dict[str, Any]) -> None:
        record = self._records.get(record_id)
        if record is None:
            raise VaporError(f'Record "{record_id}" does not exist.')
        type_def = self._schema.types[record.type]
        merged   = {**record.data, **partial}
        validate_record_data(record.type, type_def, merged)
        self._unindex_fields(record.type, record_id, record.data, list(partial.keys()))
        updated = VaporRecord(id=record_id, type=record.type, data=merged,
                              _created_at=record._created_at, _updated_at=int(time.time() * 1000))
        self._records[record_id] = updated
        self._index_fields(record.type, record_id, merged, list(partial.keys()))

    def update_where(self, type_name: str, where: dict[str, Any],
                     data: dict[str, Any]) -> int:
        """
        NEW v2.0: Bulk conditional update.
        Update all records of type_name whose fields match the where dict
        (exact equality on all where keys).

        Returns count of records updated.

        This is O(type_total) but with drastically lower Python overhead
        than calling update() N times in a loop, because:
        - Only one schema validation per type (not per record)
        - One index un-index/re-index pass rather than N passes
        - Python-side loop is tight with no function-call overhead per record

        Example:
            # Color all pixels in cluster c0042 to red in one call
            updated = record_store.update_where(
                "Pixel",
                where={"cluster": "c0042"},
                data={"r": 200.0, "g": 50.0, "b": 50.0}
            )
        """
        type_def = self._schema.types.get(type_name)
        if type_def is None:
            raise VaporError(f'Unknown type "{type_name}".')

        # Validate data fields against schema once
        # We use a dummy merged to check types — don't need a real record for this
        # Just ensure every key in data is a valid field
        for field_name in data:
            if field_name not in type_def.fields:
                # Unknown fields: stored but not indexed (same behavior as update)
                pass

        # Get all record IDs of this type
        ids_of_type = set(self._by_type.get(type_name, set()))
        if not ids_of_type:
            return 0

        updated_count = 0
        now = int(time.time() * 1000)

        for record_id in ids_of_type:
            record = self._records.get(record_id)
            if record is None:
                continue

            # Check all where conditions (exact equality)
            match = True
            for field_name, expected_val in where.items():
                actual_val = record.data.get(field_name)
                # Normalize for comparison: booleans, strings
                if isinstance(expected_val, bool):
                    if actual_val != expected_val:
                        match = False; break
                elif isinstance(expected_val, str):
                    if str(actual_val).lower() != str(expected_val).lower():
                        match = False; break
                else:
                    # Numeric: compare with small tolerance
                    try:
                        if abs(float(actual_val) - float(expected_val)) > 1e-9:
                            match = False; break
                    except (TypeError, ValueError):
                        if actual_val != expected_val:
                            match = False; break

            if not match:
                continue

            # Apply update
            merged = {**record.data, **data}
            self._unindex_fields(record.type, record_id, record.data, list(data.keys()))
            updated_record = VaporRecord(
                id=record_id, type=record.type, data=merged,
                _created_at=record._created_at, _updated_at=now
            )
            self._records[record_id] = updated_record
            self._index_fields(record.type, record_id, merged, list(data.keys()))
            updated_count += 1

        return updated_count

    def delete(self, record_id: str) -> None:
        record = self._records.pop(record_id, None)
        if record is None: return
        self._unindex_record(record.type, record_id, record.data)
        self._by_type.get(record.type, set()).discard(record_id)

    def get(self, record_id: str) -> VaporRecord | None:
        return self._records.get(record_id)

    def has(self, record_id: str) -> bool:
        return record_id in self._records

    def get_all(self) -> list[VaporRecord]:
        return list(self._records.values())

    def get_all_by_type(self, type_name: str) -> list[VaporRecord]:
        ids = self._by_type.get(type_name, set())
        return [self._records[i] for i in ids if i in self._records]

    def get_type_id_set(self, type_name: str) -> set[str]:
        return set(self._by_type.get(type_name, set()))

    def get_indexes(self, type_name: str) -> _TypeIndexes | None:
        return self._type_idxs.get(type_name)

    def get_types(self) -> list[str]:
        return list(self._by_type.keys())

    def _index_record(self, type_name: str, record_id: str, data: dict) -> None:
        self._index_fields(type_name, record_id, data, list(data.keys()))

    def _index_fields(self, type_name: str, record_id: str, data: dict, fields: list[str]) -> None:
        ti = self._type_idxs[type_name]
        for field_name in fields:
            value    = data.get(field_name)
            strategy = ti.strategies.get(field_name, "none")
            if strategy == "none": continue
            if strategy == "exact":   ti.exact.add(field_name, value, record_id)
            elif strategy == "keyword": ti.keyword.add(field_name, value, record_id)
            elif strategy == "prefix":  ti.prefix.add(field_name, value, record_id)
            elif strategy == "range":   ti.range_.add(field_name, value, record_id)

    def _unindex_record(self, type_name: str, record_id: str, data: dict) -> None:
        self._unindex_fields(type_name, record_id, data, list(data.keys()))

    def _unindex_fields(self, type_name: str, record_id: str, data: dict, fields: list[str]) -> None:
        ti = self._type_idxs[type_name]
        for field_name in fields:
            value    = data.get(field_name)
            strategy = ti.strategies.get(field_name, "none")
            if strategy == "none": continue
            if strategy == "exact":   ti.exact.remove(field_name, value, record_id)
            elif strategy == "keyword": ti.keyword.remove(record_id)
            elif strategy == "prefix":  ti.prefix.remove(field_name, value, record_id)
            elif strategy == "range":   ti.range_.remove(field_name, value, record_id)

    @property
    def total_records(self) -> int: return len(self._records)

    @property
    def record_counts_by_type(self) -> dict[str, int]:
        return {t: len(ids) for t, ids in self._by_type.items()}

    @property
    def index_stats(self) -> dict:
        exact = keyword = prefix = range_ = 0
        for ti in self._type_idxs.values():
            exact   += ti.exact.entry_count
            keyword += ti.keyword.token_count
            prefix  += ti.prefix.node_count
            range_  += ti.range_.entry_count
        return {"exact_entries": exact, "keyword_tokens": keyword,
                "prefix_nodes": prefix, "range_entries": range_}

    def clear(self) -> None:
        self._records.clear()
        for ti in self._type_idxs.values():
            ti.exact.clear(); ti.keyword.clear()
            ti.prefix.clear(); ti.range_.clear()
        for s in self._by_type.values(): s.clear()

    def _next_id(self) -> str:
        self._counter += 1
        return f"vpr_{int(time.time()*1000):x}_{self._counter:x}"
